arch/src/x86_64/mod.rs

```diff
@@ -869,9 +869,58 @@ pub fn configure_vcpu(
 
 /// Returns a Vec of the valid memory addresses.
 /// These should be used to configure the GuestMemory structure for the platform.
-/// For x86_64 all addresses are valid from the start of the kernel except a
-/// carve out at the end of 32bit address space.
-pub fn arch_memory_regions() -> Vec<(GuestAddress, usize, RegionType)> {
+pub fn arch_memory_regions(cpu_vendor: CpuVendor) -> Vec<(GuestAddress, usize, RegionType)> {
+    if matches!(cpu_vendor, CpuVendor::AMD) {
+        arch_memory_regions_amd()
+    } else {
+        arch_memory_regions_default()
+    }
+}
+
+pub fn arch_memory_regions_amd() -> Vec<(GuestAddress, usize, RegionType)> {
+    vec![
+        // 0 GiB ~ 3GiB: memory before the gap
+        (
+            GuestAddress(0),
+            layout::MEM_32BIT_RESERVED_START.raw_value() as usize,
```

Comment:
If your guest has less than 3GiB of RAM this would not work right? This would produce broken e820 map.

Reply:
Sorry @jinankjain, I don't think that I follow.

I fairly sure that after these regions are further processed by `MemoryManager::create_memory_regions_from_zones` and `generate_ram_ranges` function, a memory region <3 GiB region would be shrunk down to the correct size in the E820 map. 

Also, the AMD memory map in this PR is very similar to the intel memory map, especially for small guests with < 3GiB of RAM. Check out  `arch_memory_regions_default` to compare.

Reply:
Ahh right I got confused with the previous implementation of this function. It was recently changed to adapt for other scenarios.